Show pull progress in terminal for inflight pull requests

aaronlehmann opened the pull request:
This PR incorporates @samsabed's work from #12874. It consists of his original commit, a merge to bring it up to date with master, and two cleanup commits written by me.

The text from the original PR is:

Show pull progress when another client is also performing a same image pull request
This PR does it only on the pulled image level, and not layers, push

Fixes #8385

My additional commits:
- Bring this code up to date with a merge (mostly moving code to the new pull_v1.go and pull_v2.go files)
- Change poolAdd not to return both an error and other information at the same time
- Document and clean up ProgressStatus

Commit:
Change poolAdd to return a boolean instead of an error

Previously, its other return value was used even when it returned an
error. This is awkward and goes against the convention. It also could
have resulted in a nil pointer dereference when an error was returned
because of an unknown pool type. This changes the unknown pool type
error to a panic (since the pool types are hardcoded at call sites and
must always be "push" or "pull"), and returns a "found" boolean instead
of an error.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Clean up ProgressStatus

- Rename to Broadcaster

- Document exported types

- Change Wait function to just wait. Writing a message to the writer and
  adding the writer to the observers list are now handled by separate
  function calls.

- Avoid importing logrus (the condition where it was used should never
  happen, anyway).

- Make writes non-blocking

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

aaronlehmann commented:
@docker/distribution-maintainers PTAL

I have two questions about the code here that I'm hoping people can clarify for me.

- @samsabed's original PR changes poolAdd to first check both pushingPool and pullingPool regardless of whether `kind` is push or pull. For example, if we're trying to push something, and there's a matching pull happening, we'll return the ProgressStatus from the pull and wait on that instead of doing the push. This seems strange to me, and I'm wondering if it's intentional or not.

- In some places poolRemove is only called in the case where poolAdd added a new item to the pool. In other places, it's called whether or not a new item was added (after waiting for the operation to complete). This seems like it's probably harmless, because duplicate removals won't do anything, but I wonder if it's racy to do multiple calls to poolRemove (i.e. what if a new download starts before one of the poolRemove calls?). I also think it would be better for code clarity to only call poolRemove once for each item in the pool. Again, I'm wondering if this is intentional or not.

dmp42 commented:
cc @docker/distribution-maintainers and @cpuguy83

stevvooe commented on pkg/progressreader/progressstatus.go:
```diff
@@ -0,0 +1,78 @@
+// ProgressStatus keeps track of one or more observers watching the progress
+// of an operation. For example, if multiple clients are trying to pull an
+// image, they share a ProgressStatus for the download operation.
+type ProgressStatus struct {
```
This is more of a broadcaster, than anything else. This isn't really a `ProgressStatus`, which kind of confuses the point of this struct.

aaronlehmann replied:
How about `Broadcaster`? Since it's in the `progressreader` namespace, `progressreader.Broadcaster` should make it pretty clear what it's for.

stevvooe replied:
Sounds good.

stevvooe commented on pkg/progressreader/progressstatus.go:
```diff
@@ -0,0 +1,78 @@
+	ps.history.Write(p)
+	for _, w := range ps.observers {
+		// copy paste from MultiWriter, replaced return with continue
+		n, err = w.Write(p)
```
This will block all writers on the slowest writer or, worse, one that blocks and does not proceed.

aaronlehmann replied:
Would it be better to buffer data that writers can't keep up with, possibly indefinitely?

stevvooe replied:
That is not a tenable option, either. This was part of my original issue with this PR.

I got around this in the [`notifications` package](https://github.com/docker/distribution/blob/master/notifications/sinks.go#L83) (Note that that particular package is for `Events`, but the principles apply to byte buffers) by requiring that each writer added was non-blocking. This was accomplished by creating write queues. Note that that particular package is for `Events`, but the principles apply to byte buffers.

That is a lot more work and hard to get correct, but the only way to ensure that bad writers don't block other writers.

stevvooe commented on pkg/progressreader/progressstatus.go:
```diff
@@ -0,0 +1,78 @@
+
+// AddObserver adds an observer to the ProgressStatus. The new observer
+// receives the data from the history buffer, and also all subsequent data.
+func (ps *ProgressStatus) AddObserver(w io.Writer) {
```
`Add` is sufficient for this method name.

stevvooe commented on pkg/progressreader/progressstatus.go:
```diff
@@ -0,0 +1,78 @@
+}
+
+// Done signals to all observers that the operation has finished.
+func (ps *ProgressStatus) Done() {
```
Should this just be a `Close` method?

stevvooe commented on pkg/progressreader/progressstatus.go:
```diff
@@ -0,0 +1,78 @@
+// Done signals to all observers that the operation has finished.
+func (ps *ProgressStatus) Done() {
+	ps.Lock()
+	close(ps.c)
```
This should probably also check for flushers. I am not sure what to do with `WriteClosers`, if we have them.

endophage commented on graph/tags.go:
```diff
@@ -427,27 +428,32 @@ func validateDigest(dgst string) error {
+// poolAdd checks if a push or pull is already running, and returns (ps, true)
+// if a running operation is found. Otherwise, it creates a new one and returns
+// (ps, false).
+func (store *TagStore) poolAdd(kind, key string) (*progressreader.ProgressStatus, bool) {
```
Given we always know if it's a push or pull, it feels odd to have that information passed in an argument. Seems like there should be methods like `pushAdd` and `pullAdd` or something similar.

stevvooe replied:
Let's make an explicit pool object, rather than pollute the method space of `TagStore`.

stevvooe commented on pkg/progressreader/progressstatus.go:
```diff
@@ -0,0 +1,78 @@
+
+// Wait blocks until the operation is marked as completed by the Done method.
+func (ps *ProgressStatus) Wait() error {
+	ps.Lock()
```
This can simply be the following:

```
<-ps.c
return nil
```

There is no need to lock it. The field isn't (and shouldn't) be written concurrently.

aaronlehmann replied:
Agreed. Originally this method did some other things, but now the lock isn't necessary.

endophage commented on graph/tags.go:
```diff
@@ -427,27 +428,32 @@ func validateDigest(dgst string) error {
 
-	if c, exists := store.pullingPool[key]; exists {
-		return c, fmt.Errorf("pull %s is already in progress", key)
+	if p, exists := store.pullingPool[key]; exists {
```
I know @aaronlehmann mentioned this is the general comments, I would also like to better understand why we would check for a key in the pool for the opposing action (i.e. check a "pull" key in the push pool).

stevvooe commented on graph/pull_v2.go:
```diff
@@ -193,9 +191,8 @@ func (p *v2Puller) download(di *downloadInfo) {
 }
 
-func (p *v2Puller) pullV2Tag(tag, taggedName string) (verified bool, err error) {
+func (p *v2Puller) pullV2Tag(out io.Writer, tag, taggedName string) (verified bool, err error) {
```
`out` should already be present on `v2puller`.

aaronlehmann replied:
I added this argument so that `pullV2Tag` can write to the `ProgressStatus` writer, not its usual writer.

aaronlehmann replied:
Er, I'm missing an empty "default" case, but that was the intent.

endophage commented on pkg/progressreader/progressstatus.go:
```diff
@@ -0,0 +1,78 @@
+	ps.Lock()
+	defer ps.Unlock()
+	ps.history.Write(p)
+	for _, w := range ps.observers {
```
Is it possible for a lot of/slow observers to cause the lock to be held for a long time while we write to all of them? Seems like it would be better to explicity Unlock after ps.history.Write rather than holding the lock until the end of the method.

aaronlehmann replied:
We need to hold the lock while iterating over `ps.observers`. It would be possible to make a copy of this slice and then release the lock, but I'm not sure blocking here is really harmful.

cpuguy83 commented:
A lot o this writer logic is already in https://github.com/docker/docker/blob/master/pkg/broadcastwriter/broadcastwriter.go.

stevvooe commented:
@cpuguy83 I knew I had seen this pattern elsewhere. Thanks for the tip.

Looks like the one in this PR supports historical joining, which is necessary for progress output. Also, `BroadcastWriter` has the same issue with misbehaved writers.

Looks like there is also a bug here in the return: 'https://github.com/docker/docker/blob/master/pkg/broadcastwriter/broadcastwriter.go#L32'. Write doesn't necessary always write all of `p`. This method should really accept the write and queue it up for each writer.

cpuguy83 commented:
@stevvooe sort of, but a short write should produce an error, and as such the writer would be evicted.

stevvooe commented:
@cpuguy83 What about a write that blocks forever?

cpuguy83 commented:
Yes, that's a problem.

aaronlehmann commented:
I pushed some changes:

- Rename ProgressStatus to Broadcaster
- Rename some methods according to @stevvooe 
- Make writes non-blocking, by spawning a goroutine for each `io.Writer`. The goroutine waits for a signal that new data is available, and reads it out of the history buffer, holding after acquiring a read lock.

I'd still like to address the `poolAdd` issues.

cpuguy83 commented on pkg/progressreader/broadcaster.go:
```diff
@@ -0,0 +1,98 @@
+	// goroutine to wake up.
+	for _, dataSignal := range broadcaster.dataSignals {
+		select {
+		case dataSignal <- struct{}{}:
```
This isn't really non-blocking and doesn't solve the blocked writer issue.

aaronlehmann replied:
A blocked writer will not block this function. The write to dataSignal is non-blocking because of the select statement.

aaronlehmann replied:
Er, I'm missing an empty "default" case, but that was the intent.

stevvooe replied:
This might be a use case for [`sync.Cond`](http://golang.org/pkg/sync/#Cond), over channels.

In the `(*Broadcaster).Write`:

```
	broadcaster.Lock()
	defer broadcaster.Unlock()

	broadcaster.history.Write(p)
	broadcaster.cond.Broadcast()
```

In `(*Broadcaster). receiveWrites`:

```
for {
	broadcaster.cond.Wait()
	// ... catch up with write.

	// ... close detection
	select {
	case <-broadcast.c:
		return
	default:
		// noop: wait for write.
	}
}
```

Note that you can still do close detection using channels. Perhaps, we need a `(*Broadcaster).done`

aaronlehmann replied:
I considered using `sync.Cond`, but liked the for-range loop structure that channels allowed. I think what you're suggesting is very clean, though. I'll give that a try.

stevvooe commented on pkg/progressreader/broadcaster.go:
```diff
@@ -0,0 +1,98 @@
+// Write adds data to the history buffer, and also writes it to all current
+// observers.
+func (broadcaster *Broadcaster) Write(p []byte) (n int, err error) {
+	broadcaster.Lock()
```
This should fail if the broadcaster is closed.

stevvooe commented on pkg/progressreader/broadcaster.go:
```diff
@@ -0,0 +1,98 @@
+// Close signals to all observers that the operation has finished.
+func (broadcaster *Broadcaster) Close() {
+	broadcaster.Lock()
+	close(broadcaster.c)
```
This needs to close the broadcaster for business and flush all outstanding writes.

cpuguy83 replied:
sync.Once here for safety?

aaronlehmann replied:
A followup patch that I'm working for the layer pull race on makes it possible to safely call `Close` twice. It shouldn't be an issue for this PR because `Close` is only called once, in `poolRemove`. But I'm happy to include that part of the change in this PR if you'd like. This is the diff:

```go
diff --git a/pkg/progressreader/broadcaster.go b/pkg/progressreader/broadcaster.go
index 088d4a1..4b08ce4 100644
--- a/pkg/progressreader/broadcaster.go
+++ b/pkg/progressreader/broadcaster.go
@@ -23,6 +23,9 @@ type Broadcaster struct {
        history bytes.Buffer
        // wg is a WaitGroup used to wait for all writes to finish on Close
        wg sync.WaitGroup
+       // isClosed is set to true when Close is called to avoid closing c
+       // multiple times.
+       isClosed bool
 }
 
 // NewBroadcaster returns a Broadcaster structure
@@ -124,6 +127,11 @@ func (broadcaster *Broadcaster) Add(w io.Writer) error {
 // Close signals to all observers that the operation has finished.
 func (broadcaster *Broadcaster) Close() {
        broadcaster.Lock()
+       if broadcaster.isClosed {
+               broadcaster.Unlock()
+               return
+       }
+       broadcaster.isClosed = true
        close(broadcaster.c)
        broadcaster.cond.Broadcast()
        broadcaster.Unlock()
```

cpuguy83 replied:
Yeah, I'd prefer to fix here b/c this could cause a panic.
Any reason to use the bool vs sync.Once (curious)?

aaronlehmann replied:
Repushed with this change included.

I'm using a simple bool instead of `sync.Once` because the `Close` function already has to take a lock, so it doesn't need the additional synchronization that `sync.Once` provides.

aaronlehmann commented:
I've updated `Broadcaster` to address the feedback. It now uses a condition variable to signal to goroutines that new data is available. It also uses a waitgroup to make `Close` block until all writes complete. Methods will return errors if they are called on a closed `Broadcaster`.

Note that this is NOT ready to merge. I still need to figure out the answers to the questions I asked at the start of this PR, and address the `poolAdd` feedback.

aaronlehmann commented:
I asked earlier:

> @samsabed's original PR changes poolAdd to first check both pushingPool and pullingPool regardless of whether kind is push or pull. For example, if we're trying to push something, and there's a matching pull happening, we'll return the ProgressStatus from the pull and wait on that instead of doing the push. This seems strange to me, and I'm wondering if it's intentional or not.

Looking into this some more, the existing code (before either PR) returned an error if either a push or pull was in progress, when trying to start a new push or pull for a tag. I expect this means it's not safe for a concurrent push and pull operation to happen on the same tag, unless someone knows otherwise.

So I think the best action to take if a pull is attempted while a push is in progress (or vice versa) is to wait for the first operation to finish, and then try again. Does this make sense?

This would mean `poolAdd` should return extra a specific value when an operation of a different type is already in progress, and the caller would handle this by waiting for that operation (and displaying its progress output), then retrying when it finishes. It would also mean poolAdd should stay part of the TagStore namespace instead of being moved to a separate pool object, so it can continue to check both `pullingPool` and `pushingPool`.

I'd really appreciate any feedback/insight on this. I want to make sure this PR is moving in the right direction.

aaronlehmann commented:
I actually think this is ready to merge as is.

Trying to start a push while a pull of the same thing is running will fail. This makes sense because something can't be pushed while it's still being pulled.

Trying to start a pull while a push of the same thing is running will wait for the push to complete, and not start an actual pull. This seemed wrong to me at first, but the pull would be a no-op if we did it immediately after pushing. So I think the behavior is reasonable.

I ran into a race while testing this, but I think it's best to fix that in a separate PR, since it's not directly related to this (I can also reproduce it on master). I sometimes see concurrent pulls fail like this:

    Pulling repository docker.io/library/golang
    7b9d831c9cf1: Download complete 
    could not find image: no such id: 7b9d831c9cf1b23ebec65c9293e77c7266b4904602fe86ca6b5ee74bace734b9

This is because waits on downloads only wait until the download itself finish, not until the layer is registered in the graph and usable. There's actually a FIXME about this in pull_v2.go:

    // FIXME: Pool release here for parallel tag pull (ensures any downloads block until fully extracted)

I'd appreciate any testing people can do on this PR, or suggestions for how to make useful automated tests. This seems like a tricky situation for automated tests because it involves concurrency.

aaronlehmann commented:
Keep in mind if testing: this PR depends on #15728 (now in master) to work properly.

dmcgowan commented on graph/pull_v2.go:
```diff
@@ -121,18 +120,17 @@ func (p *v2Puller) download(di *downloadInfo) {
-		}
+	broadcaster, found := p.poolAdd("pull", "img:"+di.img.ID)
+	if found {
+		out.Write(p.sf.FormatProgress(stringid.TruncateID(di.img.ID), "Layer already being pulled by another client. Waiting.", nil))
```
These messages no longer get seen, I think we can just remove them

dmcgowan commented on graph/pull_v2.go:
```diff
@@ -73,30 +73,29 @@ func (p *v2Puller) pullV2Repository(tag string) (err error) {
+	broadcaster, found := p.poolAdd("pull", taggedName)
+	if found {
+		// Another pull of the same repository is already taking place; just wait for it to finish
+		p.config.OutStream.Write(p.sf.FormatStatus("", "Repository %s already being pulled by another client. Waiting.", p.repoInfo.LocalName))
```
V1 does not have an equivalent message, but I am questioning whether this is really necessary given the following output is from the pull.

Commit:
Remove "...already being pulled" messages

These don't get seen, so they are unnecessary.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

aaronlehmann commented:
Removed the "...already being pulled" messages.

Commit:
Show pull progress in terminal for inflight pull requests

Based on #12874 from Sam Abed <sam.abed@gmail.com>. His original commit
was brought up to date by manually porting the changes in pull.go into
the new code in pull_v1.go and pull_v2.go.

Fixes #8385

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

dmcgowan commented on pkg/progressreader/broadcaster.go:
```diff
@@ -0,0 +1,133 @@
+		}
+
+		// detect closure of the broadcast writer
+		if broadcaster.closed() {
```
I am wondering if this should be done inside the lock. I am thinking it is possible the Close is called after this check and before the lock is acquired, missing the broadcast and waiting forever.

aaronlehmann replied:
Very good catch. I changed this code to check for the closure after the lock is taken, and I also made it skip the condition variable wait if there's more data available after taking the lock (which is a similar, but less severe problem).

dmcgowan commented on pkg/progressreader/broadcaster.go:
```diff
@@ -0,0 +1,137 @@
+			return
+		}
+
+		if len(broadcaster.history.Bytes()) == n {
```
`len(broadcaster.history.Bytes())` is equivalent to `broadcaster.history.Len()`

http://golang.org/pkg/bytes/#Buffer.Len

dmcgowan commented:
LGTM

aaronlehmann commented:
BTW, the race fix that I mentioned is here: https://github.com/aaronlehmann/docker/commit/a6a4d61cd725083bece9a47403c34497522198d8

I'll make a pull request once this one is merged.

stevvooe commented:
LGTM

aaronlehmann commented:
Rebased to fix merge conflicts. Also, fixed an issue with Broadcaster's `receiveWrites` method. If this function returned due to Write failing, it needed to call `wg.Done` - otherwise the `Close` method will block forever.

aaronlehmann commented:
TestPushInterrupt is failing in janky, but it's passing locally for me.

I see that the timeout was just changed in 349aef06d51a9425f70a877585fe45283366c2c4. I think the failure isn't related to this commit.

cpuguy83 commented:
Well I was testing this manually, but there's a hub outage.

stevvooe commented:
@aaronlehmann That test relies on a race condition. I was trying to find a value that works on a local machine and jenkins, but we should probably just get rid of it.

cpuguy83 commented:
Of course this is breaking docker-py tests

aaronlehmann commented:
> Of course this is breaking docker-py tests

Confirmed locally. I'll look into this tomorrow.

Do you think those tests are sensitive to `docker pull` output, or are they indicating something is actually broken?

aaronlehmann commented:
The root cause of the failures is that one test is apparently receiving invalid JSON from the pull API endpoint. That's strange because I can't think of any way this PR would affect JSON encoding, nor any obvious way the Broadcaster could corrupt this JSON. The issue might be the Python client expecting the HTTP stream to be flushed for each JSON message that it carries, but `Broadcaster` doesn't do this flushing. I'll take a closer look tomorrow.

Commit:
Make the broadcaster write messages to the observers in the same units they were written to the broadcaster

This means the writing to a WriteFlusher will flush in the same places
as it would if the broadcaster wasn't sitting in front of it.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

aaronlehmann commented:
> The issue might be the Python client expecting the HTTP stream to be flushed for each JSON message that it carries, but Broadcaster doesn't do this flushing.

That was indeed the problem. I've changed `Broadcaster` to store a sequence of individual messages, and write those to observers in the same increments.

It's a bit unfortunate that WriteFlusher overloads `Write` to mean "write and flush". I think this is a confusing API decision that could lead to more problems in the future. Oh well.

aaronlehmann commented:
Tests look good now. The `TestPushInterrupt` failure is unrelated. I see that failing for other PRs.

icecrime commented:
LGTM
